### PR TITLE
feat: Use blake2b as the hash function uniformly

### PR DESCRIPTION
**Pull request by zhangsoledad**
https://github.com/nervosnetwork/ckb/issues/286

- [x] util/merkle-tree https://github.com/nervosnetwork/merkle-tree/pull/8
- [x] mruby-contracts https://github.com/nervosnetwork/mruby-contracts/pull/5
- [x] ruby sdk https://github.com/nervosnetwork/ckb-demo-ruby-sdk/pull/66
- [x] swift sdk https://github.com/nervosnetwork/ckb-sdk-swift/pull/29
- [x] java-sdk https://github.com/nervosnetwork/ckb-sdk-java/pull/24
- [x] js-sdk https://github.com/nervosnetwork/ckb-sdk-js/pull/56

### CODE REVIEW

**Comment by doitian** [comment]
Any plan to update SDKs?

cc @ashchan

File: chain/src/tests/find_fork.rs [hunk 1]
```diff
@@ -102,14 +102,14 @@ fn test_find_fork_case2() {
 
     let mut parent = genesis.clone();
     for i in 0..4 {
-        let new_block = gen_block(&parent, i, U256::from(100u64), vec![], vec![]);
+        let new_block = gen_block(&parent, i + 1, U256::from(100u64), vec![], vec![]);
```
Comment (doitian) [inline chain/src/tests/find_fork.rs]: Why `i` and `i + 1` have changed their positions here and in line 112?

**Comment by doitian** [comment]
It seems that the merkle tree merge function has not been changed to blake2b yet.

File: chain/src/tests/find_fork.rs (hunk 1 above)
Comment (ashchan) [reply chain/src/tests/find_fork.rs]: SDKs could catch up soon. But I want to know which version would include this change, v0.7.0?

Comment (doitian) [reply chain/src/tests/find_fork.rs]: Not decided yet, depend on you.

**Comment by ashchan** [comment]
Then will plan and start soon. 

iPhoneから送信

2019/03/06 0:16、ian <notifications@github.com>のメール:

File: script/src/syscalls/mod.rs [hunk 2]
```diff
@@ -738,7 +738,7 @@ mod tests {
         machine.registers_mut()[A7] = LOAD_INPUT_BY_FIELD_SYSCALL_NUMBER; // syscall number
 
         let unlock = Script::new(0, vec![], None, Some(vec![]), vec![]);
-        let sha3_data = sha3_256(data);
+        let sha3_data = blake2b_256(data);
```
Comment (ashchan) [inline script/src/syscalls/mod.rs]: Why still name the var `sha3_data`?

File: core/src/script.rs [hunk 3]
```diff
@@ -167,7 +167,7 @@ impl Script {
                 for argument in &self.signed_args {
                     bytes.write_all(argument).unwrap();
                 }
-                sha3_256(bytes).into()
+                blake2b_256(bytes).into()
```
Comment (ashchan) [inline core/src/script.rs]: Can we have a simple unit test for `Script.type_hash`?

File: chain/src/tests/find_fork.rs (hunk 1 above)
Comment (zhangsoledad) [reply chain/src/tests/find_fork.rs]: I... will rewrite this part.
Should not depend on the hash comparison.
@doitian

**Comment by zhangsoledad** [comment]
https://github.com/nervosnetwork/merkle-tree/pull/8

File: nodes_template/default.toml [hunk 4]
```diff
@@ -44,4 +44,4 @@ trace = 100
 txs_verify_cache_size = 100000
 
 [block_assembler]
-type_hash = "0x0da2fe99fe549e082d4ed483c2e968a89ea8d11aabf5d79e5cbf06522de6e674"
+type_hash = "0x8954a4ac5e5c33eb7aa8bb91e0a000179708157729859bd8cf7e2278e1e12980"
```
Comment (quake) [inline nodes_template/default.toml]: is it a type hash for `always_success` script?

Comment (zhangsoledad) [reply nodes_template/default.toml]: yes

Comment (quake) [reply nodes_template/default.toml]: Please change this type hash also, and add a comment in default.toml:
https://github.com/nervosnetwork/ckb/blob/develop/test/fixtures/nodes_template/default.toml#L47

**Comment by ashchan** [comment]
All the related SDKs (`rc/v0.7.0` branch) and scripts (`master` branch) repositories are reviewed and merged.

**Pull request merged**